contracts/inventory/AchievementFacet.sol

```diff
@@ -0,0 +1,307 @@
+    }
+
+    // Revokes the admin role from an account.
+    function revokeAdminPrivilege(address admin) external {
+        LibDiamondMoonstream.enforceIsContractOwner();
+
+        LibAchievement.AchievementStorage storage acs = LibAchievement
+            .achievementStorage();
+
+        ITerminus terminus = ITerminus(address(this));
+
+        uint256 balance = terminus.balanceOf(admin, acs.adminTerminusPoolID);
+        if (balance > 0) {
+            terminus.burn(admin, acs.adminTerminusPoolID, balance);
+        }
+
+        emit AdminPrivilegeRevoked(admin);
+    }
+
+    function createSlot(bool, string memory) public override returns (uint256) {
+        revert("This method is disabled");
+    }
+
+    function markItemAsEquippableInSlot(
+        uint256,
+        uint256,
+        address,
+        uint256,
+        uint256
+    ) public override {
+        revert("This method is disabled");
+    }
+
+    function createAchievementSlot(
+        string memory metadataURI
+    ) public returns (uint256) {
+        _enforceIsAdmin();
+
+        LibInventory.InventoryStorage storage istore = LibInventory
+            .inventoryStorage();
+
+        // Slots are 1-indexed!
+        istore.NumSlots += 1;
+        uint256 newSlot = istore.NumSlots;
+        // save the slot type!
+        istore.SlotData[newSlot] = Slot({
+            SlotURI: metadataURI,
+            SlotIsPersistent: true
+        });
+
+        // create terminus token
+        ITerminus terminus = ITerminus(address(this));
+        uint256 poolId = terminus.createPoolV2(
+            type(uint256).max,
+            false,
+            true,
+            metadataURI
+        );
+
+        require(
+            newSlot == poolId,
+            "AchievementFacet.createAchievementSlot: Slot id and pool id are mismatched."
+        );
+
+        // mark token equippable in slot
+        istore.SlotEligibleItems[newSlot][LibInventory.ERC1155_ITEM_TYPE][
+            address(this)
+        ][poolId] = 1;
+
+        emit ItemMarkedAsEquippableInSlot(
+            newSlot,
+            LibInventory.ERC1155_ITEM_TYPE,
+            address(this),
+            poolId,
+            1
+        );
+
+        emit SlotCreated(msg.sender, newSlot);
+        emit NewSlotURI(newSlot);
+        emit NewSlotPersistence(newSlot, true);
+
+        return newSlot;
+    }
+
+    function _mintToInventory(uint256 subjectTokenId, uint256 poolId) internal {
+        LibInventory.InventoryStorage storage istore = LibInventory
+            .inventoryStorage();
+
+        // We could also revert the transaction and say the subject already has this achievement. Or we
+        // could do nothing and assume the achievement is correctly appplied already.
+        if (
+            istore
+            .EquippedItems[istore.ContractERC721Address][subjectTokenId][poolId]
+                .ItemType != 0
+        ) {
+            _unequip(subjectTokenId, poolId, true, 0);
```

Comment:
Since the slots are persistent, wouldn't this revert anyway? Or is the behavior different for the internal `_unequip` method? Or different for players and admins?

There's a lot of reasoning that goes into understanding what happens here.

I'd rather we had an explicit reversion. Perhaps we need a method that helps us rescue ourselves from erroneously granting achievements?